src/split_single_dyn_dim.cpp

```diff
@@ -0,0 +1,137 @@
+/*
+ * The MIT License (MIT)
+ *
+ * Copyright (c) 2015-2023 Advanced Micro Devices, Inc. All rights reserved.
+ *
+ * Permission is hereby granted, free of charge, to any person obtaining a copy
+ * of this software and associated documentation files (the "Software"), to deal
+ * in the Software without restriction, including without limitation the rights
+ * to use, copy, modify, merge, publish, distribute, sublicense, and/or sell
+ * copies of the Software, and to permit persons to whom the Software is
+ * furnished to do so, subject to the following conditions:
+ *
+ * The above copyright notice and this permission notice shall be included in
+ * all copies or substantial portions of the Software.
+ *
+ * THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, EXPRESS OR
+ * IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY,
+ * FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT.  IN NO EVENT SHALL THE
+ * AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER
+ * LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM,
+ * OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN
+ * THE SOFTWARE.
+ */
+
+#include <migraphx/split_single_dyn_dim.hpp>
+#include <migraphx/module.hpp>
+#include <migraphx/pass_manager.hpp>
+#include <migraphx/functional.hpp>
+#include <migraphx/make_op.hpp>
+
+namespace migraphx {
+inline namespace MIGRAPHX_INLINE_NS {
+
+bool has_one_dyn_dim(const std::unordered_map<std::string, shape>& param_shapes,
+                     std::string& dyn_param_str,
+                     size_t& dyn_index,
+                     size_t& min_dim,
+                     size_t& max_dim)
+{
+    // True if parameters contain exactly one dynamic shape with exactly one non-fixed
+    // dynamic_dimension.
+    size_t num_dynamic = 0;
+    for(const auto& ps : param_shapes)
+    {
+        if(ps.second.dynamic())
+        {
+            num_dynamic += 1;
+            if(num_dynamic > 1)
+            {
+                return false;
+            }
+            size_t num_nf = 0;
+            auto dds      = ps.second.dyn_dims();
+            for(size_t i = 0; i < dds.size(); ++i)
```

Comment:
Use `find_if` for this loop.

Reply:
Same issue as above. I think I can use `accumulate` instead?

Reply:
No need to use `std::accumulate`, just do a second `find_if` call:

```cpp
auto not_fixed = [](const auto& dd) {
    return not dd.is_fixed();
};
auto dd_it = std::find_if(dds.begin(), dds.end(), not_fixed);
// Maybe this should be an assert?
if (dd_it == dds.end())
    return std::nullopt;
// Check for second non-fixed dimension
if (std::find_if(dd_it, dds.end(), not_fixed) != param_shapes.end())
    return std::nullopt;
result.max = dd_it->max;
result.min = dd_it->min;
result.index = dd_it - dds.begin();
return result;
```

Reply:
Also `any_of` can be used here as well for the second search.